Refactor Forest

serprex opened the pull request:
Vanilla was already done, MQ only needed a bit of work around Forest's useless corridors

Addressed region name bikeshedding from #5318

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4877877505.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4877891494.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4878377288.zip)
<!--- section:artifacts:end -->

serprex committed:
refactor forest

serprex commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -404,41 +422,64 @@ void RegionTable_Init_ForestTemple() {
-        Entrance(RR_FOREST_TEMPLE_MQ_NE_OUTDOORS,         []{return (((logic->CanUse(RG_IRON_BOOTS) || logic->CanUse(RG_LONGSHOT) || (ctx->GetTrickOption(RT_FOREST_MQ_WELL_SWIM) && logic->CanUse(RG_HOOKSHOT))) && logic->HasItem(RG_BRONZE_SCALE)) || logic->HasItem(RG_GOLDEN_SCALE)) && logic->WaterTimer() >= 16;}),
-        Entrance(RR_FOREST_TEMPLE_MQ_OUTDOORS_TOP_LEDGES, []{return logic->CanUse(RG_FIRE_ARROWS);}),
+        Entrance(RR_FOREST_TEMPLE_MQ_NE_COURTYARD,       []{return (((logic->CanUse(RG_IRON_BOOTS) || logic->CanUse(RG_LONGSHOT) || (ctx->GetTrickOption(RT_FOREST_MQ_WELL_SWIM) && logic->CanUse(RG_HOOKSHOT))) && logic->HasItem(RG_BRONZE_SCALE)) || logic->HasItem(RG_GOLDEN_SCALE)) && logic->WaterTimer() >= 16;}),
+        Entrance(RR_FOREST_TEMPLE_MQ_NORTH_PASSAGE,      []{return logic->CanUse(RG_FIRE_ARROWS);}), // temp flag, can't be event
```
need region for coming out behind webs, also should make this & the eye switches in lobby perm flags

#5676

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -9,14 +9,14 @@ void RegionTable_Init_ForestTemple() {
+    areaTable[RR_FOREST_TEMPLE_TREES] = Region("Forest Temple Trees", SCENE_FOREST_TEMPLE, {}, {
         //Locations
         LOCATION(RC_FOREST_TEMPLE_FIRST_ROOM_CHEST, true),
         LOCATION(RC_FOREST_TEMPLE_GS_FIRST_ROOM,    (logic->IsAdult && logic->CanUse(RG_BOMB_BAG)) || logic->CanUse(RG_FAIRY_BOW) || logic->CanUse(RG_HOOKSHOT) || logic->CanUse(RG_BOOMERANG) || logic->CanUse(RG_FAIRY_SLINGSHOT) || logic->CanUse(RG_BOMBCHU_5) || logic->CanUse(RG_DINS_FIRE) || (ctx->GetTrickOption(RT_FOREST_FIRST_GS) && (logic->CanJumpslashExceptHammer() || (logic->IsChild && logic ...
```
Complex skull kills like this should be converted to CanKillEnemy, but it looks like this one needs ED_BOMB_THROW to be split into child and adult.

Ideally all skulls would use KillEnemy or EnemyDrop but I'm not going to pick at every detail so long as the logic is clean and works.

serprex replied:
fixing specific logic should happen in a different PR, as this refactoring is focused on macro structure more than fixing things

Fixing little things while I'm in here makes sense tho ofc

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -103,75 +103,75 @@ void RegionTable_Init_ForestTemple() {
         //Locations
-        LOCATION(RC_FOREST_TEMPLE_RAISED_ISLAND_COURTYARD_CHEST, logic->CanUse(RG_HOOKSHOT) || HasAccessTo(RR_FOREST_TEMPLE_FALLING_ROOM) || (HasAccessTo(RR_FOREST_TEMPLE_NE_OUTDOORS_UPPER) && logic->IsAdult && ctx->GetTrickOption(RT_FOREST_OUTDOORS_LEDGE) && logic->CanUse(RG_HOVER_BOOTS))),
-        LOCATION(RC_FOREST_TEMPLE_GS_RAISED_ISLAND_COURTYARD,    logic->CanUse(RG_HOOKSHOT) || (ctx->GetTrickOption(RT_FOREST_OUTDOORS_EAST_GS) && logic->CanUse(RG_BOOMERANG)) || Here(RR_FOREST_TEMPLE_FALLING_ROOM, []{return logic->CanUse(RG_FAIRY_BOW) || logic->CanUse(RG_FAIRY_SLINGSHOT) || logic->CanUse(RG_DINS_FIRE) || logic->HasExplosives();})),
+        LOCATION(RC_FOREST_TEMPLE_RAISED_ISLAND_COURTYARD_CHEST, logic->CanUse(RG_HOOKSHOT) || HasAccessTo(RR_FOREST_TEMPLE_FALLING_ROOM) || (HasAccessTo(RR_FOREST_TEMPLE_NE_COURTYARD_UPPER) && logic->IsAdult && ctx->GetTrickOption(RT_FOREST_COURTYARD_LEDGE) && logic->CanUse(RG_HOVER_BOOTS))),
```
HasAccessTo here needs removing, it's an assumption that breaks in doorsanity and bad practice besides (playthrough jank), instead the ledge should be it's own region.

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -103,75 +103,75 @@ void RegionTable_Init_ForestTemple() {
-        LOCATION(RC_FOREST_TEMPLE_RAISED_ISLAND_COURTYARD_CHEST, logic->CanUse(RG_HOOKSHOT) || HasAccessTo(RR_FOREST_TEMPLE_FALLING_ROOM) || (HasAccessTo(RR_FOREST_TEMPLE_NE_OUTDOORS_UPPER) && logic->IsAdult && ctx->GetTrickOption(RT_FOREST_OUTDOORS_LEDGE) && logic->CanUse(RG_HOVER_BOOTS))),
-        LOCATION(RC_FOREST_TEMPLE_GS_RAISED_ISLAND_COURTYARD,    logic->CanUse(RG_HOOKSHOT) || (ctx->GetTrickOption(RT_FOREST_OUTDOORS_EAST_GS) && logic->CanUse(RG_BOOMERANG)) || Here(RR_FOREST_TEMPLE_FALLING_ROOM, []{return logic->CanUse(RG_FAIRY_BOW) || logic->CanUse(RG_FAIRY_SLINGSHOT) || logic->CanUse(RG_DINS_FIRE) || logic->HasExplosives();})),
+        LOCATION(RC_FOREST_TEMPLE_RAISED_ISLAND_COURTYARD_CHEST, logic->CanUse(RG_HOOKSHOT) || HasAccessTo(RR_FOREST_TEMPLE_FALLING_ROOM) || (HasAccessTo(RR_FOREST_TEMPLE_NE_COURTYARD_UPPER) && logic->IsAdult && ctx->GetTrickOption(RT_FOREST_COURTYARD_LEDGE) && logic->CanUse(RG_HOVER_BOOTS))),
+        LOCATION(RC_FOREST_TEMPLE_GS_RAISED_ISLAND_COURTYARD,    logic->CanUse(RG_HOOKSHOT) || (ctx->GetTrickOption(RT_FOREST_COURTYARD_EAST_GS) && logic->CanUse(RG_BOOMERANG)) || Here(RR_FOREST_TEMPLE_FALLING_ROOM, []{return logic->CanUse(RG_FAIRY_BOW) || logic->CanUse(RG_FAIRY_SLINGSHOT) || logic->CanUse(RG_DINS_FIRE) || logic->HasExplosives();})),
```
This skull needs to be in multiple regions to map properly, but not all can be added immediately and should instead be commented.

- From the main lower area, which is possible but precise rang using the backthrow, but would be a trick.
- From the ledge, at drop(ED_BOOMERANG)
- From the top ledge, with kill(ED_CHILD_BOMB_THROW), to accommodate killing the skull, then dropping onto the token
- Possibly from the balcony, with Kill ED_FAR and some hovers boost or bomb boost, should they kill the skull token but not the ledge

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -103,75 +103,75 @@ void RegionTable_Init_ForestTemple() {
     });
 
-    areaTable[RR_FOREST_TEMPLE_NE_OUTDOORS_LOWER] = Region("Forest Temple NE Outdoors Lower", SCENE_FOREST_TEMPLE, {
+    areaTable[RR_FOREST_TEMPLE_NE_COURTYARD_LOWER] = Region("Forest Temple NE Courtyard Lower", SCENE_FOREST_TEMPLE, {
```
This region needs splitting into 3: the main lower section, the ledge with the chest, and the upper ledge with the scarecrow, in order to properly handle logic for the skull and chest in doorsanity.

Additionally the scarecrow here is a perm flag that can be set as child, so for doorsanity needs to be an event. The scarecrow can also be called from the door's archway, which is how RT_FOREST_DOORFRAME works. it is probably best to make the arch it's own region to simplify that.

serprex replied:
doorframe shouldn't need region, let upper set event & hookshot access will follow from lower

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -103,75 +103,75 @@ void RegionTable_Init_ForestTemple() {
-        Entrance(RR_FOREST_TEMPLE_FALLING_ROOM,      []{return ctx->GetTrickOption(RT_FOREST_DOORFRAME) && logic->CanJumpslashExceptHammer() && logic->CanUse(RG_HOVER_BOOTS) && logic->CanUse(RG_SCARECROW);}),
+        Entrance(RR_FOREST_TEMPLE_NE_COURTYARD_LOWER, []{return true;}),
+        Entrance(RR_FOREST_TEMPLE_MAP_ROOM,           []{return true;}),
+        Entrance(RR_FOREST_TEMPLE_FALLING_ROOM,       []{return ctx->GetTrickOption(RT_FOREST_DOORFRAME) && logic->CanJumpslashExceptHammer() && logic->CanUse(RG_HOVER_BOOTS) && logic->CanUse(RG_SCARECROW);}),
```
I would need to test/review this trick, but there was a cautious conversion of "CanJumpslash" to "CanJumpslashExceptHammer" at one point while cleaning up some helpers, and if this is jumpslash for distance like it looks like, hammer would work fine and this was an overly cautious conversion.

Pepper0ni replied:
OK, messed around with the trick, hammer is fine.

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -103,75 +103,75 @@ void RegionTable_Init_ForestTemple() {
+        Entrance(RR_FOREST_TEMPLE_NE_COURTYARD_UPPER, []{return Here(RR_FOREST_TEMPLE_MAP_ROOM, []{return logic->CanKillEnemy(RE_BLUE_BUBBLE);});}),
     });
 
     areaTable[RR_FOREST_TEMPLE_SEWER] = Region("Forest Temple Sewer", SCENE_FOREST_TEMPLE, {}, {
```
HasAccessTo can't be used here for age logic reasons, and shouldn't for playthrough jank reasons.

This can either be handled like well checks are in MQ with the underwater requirements being part of each COURTYARD_LOWER and the drained requirements being on the balconys, or you could make a "dry well" region that balcony connects to. I would be cautious about making a full-well region due to water timers, especially with the chest being at 1 end.

Pepper0ni replied:
Looks like this got sorted in master

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -559,8 +600,14 @@ void RegionTable_Init_ForestTemple() {
+    areaTable[RR_FOREST_TEMPLE_MQ_EAST_CORRIDOR] = Region("Forest Temple MQ East Corridor", SCENE_FOREST_TEMPLE, {}, {}, {
+        //Exits
+        Entrance(RR_FOREST_TEMPLE_MQ_AMY_ROOM, []{return true;}),
+        Entrance(RR_FOREST_TEMPLE_MQ_LOBBY,    []{return true;}),
```
There's a big skultulla blocking the way here

serprex replied:
other issue with these skulltula blocked hallways is they need to be split too

serprex replied:
only blocked in vanilla

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -559,8 +600,14 @@ void RegionTable_Init_ForestTemple() {
-        Entrance(RR_FOREST_TEMPLE_MQ_CENTRAL_AREA, []{return logic->ForestTempleAmy;}),
-        Entrance(RR_FOREST_TEMPLE_MQ_FALLING_ROOM, []{return true;}),
+        Entrance(RR_FOREST_TEMPLE_MQ_EAST_CORRIDOR, []{return logic->ForestTempleAmy;}),
+        Entrance(RR_FOREST_TEMPLE_MQ_FALLING_ROOM,  []{return true;}),
```
MB, there's a 6 key door here

serprex replied:
ugh this whole torch room -> amy room logic is going to need to be untangled, but punting on that for now. Worst case doorsanity can get by not shuffling doors to twisted hallways

Pepper0ni reviewed (changes requested):
NE courtyard is a bit of a mess, so I'm not surprised there was issues there.

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -75,25 +75,25 @@ void RegionTable_Init_ForestTemple() {
-        Entrance(RR_FOREST_TEMPLE_MAP_ROOM,          []{return true;}),
-        Entrance(RR_FOREST_TEMPLE_SEWER,             []{return logic->HasItem(RG_GOLDEN_SCALE) || logic->CanUse(RG_IRON_BOOTS);}),
-        Entrance(RR_FOREST_TEMPLE_BOSS_ENTRYWAY,     []{return false;}),
+        Entrance(RR_FOREST_TEMPLE_LOBBY,              []{return logic->CanUse(RG_SONG_OF_TIME);}),
```
The SoT block appears on the other side of the door, only as adult, doesn't have collision from the inside and can't even be cleared until you are out of it anyway. There's no need to require it here. Not even visible collision applies as you can't see it until after you've already left.

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -75,25 +75,25 @@ void RegionTable_Init_ForestTemple() {
-        Entrance(RR_FOREST_TEMPLE_SEWER,             []{return logic->HasItem(RG_GOLDEN_SCALE) || logic->CanUse(RG_IRON_BOOTS);}),
-        Entrance(RR_FOREST_TEMPLE_BOSS_ENTRYWAY,     []{return false;}),
+        Entrance(RR_FOREST_TEMPLE_LOBBY,              []{return logic->CanUse(RG_SONG_OF_TIME);}),
+        Entrance(RR_FOREST_TEMPLE_NW_COURTYARD_UPPER, []{return ctx->GetTrickOption(RT_HOVER_BOOST_SIMPLE) && ctx->GetTrickOption(RT_DAMAGE_BOOST_SIMPLE) && logic->HasExplosives() && logic->CanUse(RG_HOVER_BOOTS);}),
```
A midair ground jump can also do this trick, though may be adult only

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -197,7 +197,7 @@ void RegionTable_Init_ForestTemple() {
     }, {
```
Block push room likely needs splitting up for doorsanity, but the twisted hallway almost certainly needs special handling anyway.

In particular you do not need str1 to do anything in this room if you are entering through the top, as you get easy access to the entire room.

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -75,25 +75,25 @@ void RegionTable_Init_ForestTemple() {
         Entrance(RR_FOREST_TEMPLE_NORTH_CORRIDOR, []{return true;}),
```
This door locks until the stalfos are defeated, also `LOGIC_FOREST_CLEAR_BETWEEN_JOELLE_AND_BETH` should be implemented here.

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -422,41 +440,64 @@ void RegionTable_Init_ForestTemple() {
-        Entrance(RR_FOREST_TEMPLE_MQ_OUTDOORS_TOP_LEDGES, []{return logic->CanUse(RG_FIRE_ARROWS);}),
+        Entrance(RR_FOREST_TEMPLE_MQ_NE_COURTYARD,       []{return (((logic->CanUse(RG_IRON_BOOTS) || logic->CanUse(RG_LONGSHOT) || (ctx->GetTrickOption(RT_FOREST_MQ_WELL_SWIM) && logic->CanUse(RG_HOOKSHOT))) && logic->HasItem(RG_BRONZE_SCALE)) || logic->HasItem(RG_GOLDEN_SCALE)) && logic->WaterTimer() >= 16;}),
+        Entrance(RR_FOREST_TEMPLE_MQ_NORTH_PASSAGE,      []{return logic->CanUse(RG_FIRE_ARROWS);}), // temp flag, can't be event
+        Entrance(RR_FOREST_TEMPLE_MQ_NW_COURTYARD_LEDGE, []{return ctx->GetTrickOption(RT_HOVER_BOOST_SIMPLE) && ctx->GetTrickOption(RT_DAMAGE_BOOST_SIMPLE) && logic->HasExplosives() && logic->CanUse(RG_HOVER_BOOTS);}),
```
Also needs the middair ground jump method adding

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -564,8 +605,8 @@ void RegionTable_Init_ForestTemple() {
         //Exits
-        Entrance(RR_FOREST_TEMPLE_MQ_NE_OUTDOORS_LEDGE, []{return true;}),
-        Entrance(RR_FOREST_TEMPLE_MQ_AMY_ROOM,          []{return logic->SmallKeys(SCENE_FOREST_TEMPLE, 6);}),
+        Entrance(RR_FOREST_TEMPLE_MQ_NE_COURTYARD_LEDGE, []{return true;}),
```
Similar to what i said about vanilla, we want to split things up here for doorsanity, except it's already half done here.

Pepper0ni reviewed (changes requested):
There's a bunch of room splits that still need to happen here, some non-trivial

serprex committed:
feedback

serprex commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -104,50 +105,65 @@ void RegionTable_Init_ForestTemple() {
     }, {
+        //Exits
+        Entrance(RR_FOREST_TEMPLE_NE_COURTYARD_LOWER,           []{return true;}),
+        Entrance(RR_FOREST_TEMPLE_NE_COURTYARD_SCARECROW_LEDGE, []{return logic->CanUse(RG_HOOKSHOT);}),
```
```suggestion
        Entrance(RR_FOREST_TEMPLE_NE_COURTYARD_SCARECROW_LEDGE, []{return logic->CanUse(RG_SCARECROW);}),
```

serprex committed:
more feedback

serprex committed:
MQ NE island logic

serprex committed:
no kids allowed

serprex committed:
split west corridor

serprex committed:
split up block puzzle room

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -64,36 +70,37 @@ void RegionTable_Init_ForestTemple() {
+        Entrance(RR_FOREST_TEMPLE_LOBBY,              []{return true;}),
+        Entrance(RR_FOREST_TEMPLE_NW_COURTYARD_UPPER, []{return logic->CanUse(RG_HOVER_BOOTS) && ((ctx->GetTrickOption(RT_HOVER_BOOST_SIMPLE) && ctx->GetTrickOption(RT_DAMAGE_BOOST_SIMPLE) && logic->HasExplosives()) || (ctx->GetTrickOption(RT_GROUND_JUMP_HARD) && logic->IsAdult && logic->CanGroundJump()));}),
+        Entrance(RR_FOREST_TEMPLE_MAP_ROOM,           []{return true;}),
+        Entrance(RR_FOREST_TEMPLE_SEWER,              []{return logic->HasItem(RG_GOLDEN_SCALE) || logic->CanUse(RG_IRON_BOOTS);}),
```
Doorsanity can make RT_FOREST_MQ_WELL_SWIM relevant in vanilla, also tunic logic is needed as the swim lasts long enough to spawn a drown timer (see the MQ version).

Pepper0ni commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -104,50 +111,67 @@ void RegionTable_Init_ForestTemple() {
-        EventAccess(LOGIC_DEKU_BABA_NUTS,   []{return logic->CanGetDekuBabaNuts();}),
+        EventAccess(LOGIC_DEKU_BABA_STICKS,           []{return logic->CanGetDekuBabaSticks();}),
+        EventAccess(LOGIC_DEKU_BABA_NUTS,             []{return logic->CanGetDekuBabaNuts();}),
+        EventAccess(LOGIC_FOREST_SUMMON_NE_SCARECROW, []{return ctx->GetTrickOption(RT_FOREST_DOORFRAME) && logic->CanJumpslashExceptHammer() && logic->CanUse(RG_HOVER_BOOTS) && logic->ScarecrowsSong();}),
```
I think i tested the wrong trick with hammer and this was the one that was good. Will need to doublecheck the other one

serprex replied:
you can reach this with hammer hoverboot recoil. from north balcony's eastern railing you don't need to even steer

also seems adult only

Pepper0ni reviewed (approved):
I may have messed up with the forest tricks, but that's a quick patch later

Pepe20129 commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -9,27 +9,33 @@ void RegionTable_Init_ForestTemple() {
 
-    areaTable[RR_FOREST_TEMPLE_SOUTH_CORRIDOR] = Region("Forest Temple South Corridor", SCENE_FOREST_TEMPLE, {}, {}, {
+    areaTable[RR_FOREST_TEMPLE_SOUTH_CORRIDOR_SOUTH] = Region("Forest Temple South Corridor South", SCENE_FOREST_TEMPLE, {}, {}, {
         //Exits
-        Entrance(RR_FOREST_TEMPLE_FIRST_ROOM, []{return true;}),
-        Entrance(RR_FOREST_TEMPLE_LOBBY,      []{return logic->CanPassEnemy(RE_BIG_SKULLTULA);}),
+        Entrance(RR_FOREST_TEMPLE_TREES,                []{return true;}),
+        Entrance(RR_FOREST_TEMPLE_SOUTH_CORRIDOR_NORTH, []{return logic->CanPassEnemy(RE_BIG_SKULLTULA);}),
+    });
+
+    areaTable[RR_FOREST_TEMPLE_SOUTH_CORRIDOR_NORTH] = Region("Forest Temple South Corridor North", SCENE_FOREST_TEMPLE, {}, {}, {
+        //Exits
+        Entrance(RR_FOREST_TEMPLE_SOUTH_CORRIDOR_SOUTH, []{return logic->CanPassEnemy(RE_BIG_SKULLTULA);}),
+        Entrance(RR_FOREST_TEMPLE_LOBBY,                []{return true;}),
     });
```
Not the best names but I can't think of any better ones

Pepe20129 commented on soh/soh/Enhancements/randomizer/location_access/dungeons/forest_temple.cpp:
```diff
@@ -64,36 +70,37 @@ void RegionTable_Init_ForestTemple() {
         //Events
-        EventAccess(LOGIC_FAIRY_POT, []{return true;}),
+        EventAccess(LOGIC_FAIRY_POT,                            []{return true;}),
+        EventAccess(LOGIC_FOREST_CLEAR_BETWEEN_JOELLE_AND_BETH, []{return logic->CanKillEnemy(RE_STALFOS, ED_CLOSE, true, 2);}),
```
Maybe a comment at the top of the file explaining the poe names would be helpful.

serprex committed:
comment mapping poe sister colors

serprex committed:
generalize well swim, make novice

serprex committed:
hammer works

serprex committed:
upstream bug fix

serprex committed:
Replace corridor with hallway, rename east/west hallway regions to be based on their red/blue doormats

also change south hallway to overgrown hallway, & replace south/north sides with lower/upper sides
